File: Chapter 2 Sorting/algorithms2_1/src/main/scala/ca/vgorcinschi/algorithms2_4/LinkedPQ.scala
package ca.vgorcinschi.algorithms2_4

import com.google.common.annotations.VisibleForTesting

import scala.language.postfixOps

/**
  * Ex 2.4.24 Priority queue with explicit links. Implement a priority queue using a heap-ordered binary
  * tree, but use a triply linked structure instead of an array. You will need three links per childBranch:
  * two to traverse down the tree and one to traverse up the tree. Your implementation should guarantee
  * logarithmic running time per operation, even if no maximum priority-queue size is known ahead time.
  */
class LinkedPQ[Key](implicit override protected val cmp: Ordering[_ >: Key])
  extends MaxPQ[Key] with PQNodeSupport[Key] {

  private var root: Tree[Key] = EmptyTree
  private var last: Tree[Key] = EmptyTree

  override def insert(value: Key): Unit = {
    N += 1
    root = insertHelper(root, value)
    root = swim(last)
  }

  @VisibleForTesting
  protected def insertHelper(nextTree: Tree[Key], value: Key): Branch[Key] = nextTree match {
    case EmptyTree =>
      last = Tree(value)
      last.asInstanceOf[Branch[Key]]
    case branch: Branch[Key] =>
      // compare left and right sizes to see where to go
      val leftSize = branch.left.size
      val rightSize = branch.right.size

      if (leftSize <= rightSize) {
        val newLeftBranch = insertHelper(branch.left, value)
        newLeftBranch.parent = Some(branch)
        branch.left = newLeftBranch
      } else {
        val newRightBranch = insertHelper(branch.right, value)
        newRightBranch.parent = Some(branch)
        branch.right = newRightBranch
      }
      branch
  }

  def removeLastBranch(currentTree: Tree[T]): (Tree[T], T) = {
    def shortenLeftBranch: (Tree[T], T) = {
      val (updatedLeftBranch, removedElem) = removeLastBranch(currentTree.left)
      if (updatedLeftBranch != EmptyTree) {
        (Tree(currentTree.elem,
          Tree(updatedLeftBranch.elem, updatedLeftBranch.left, updatedLeftBranch.right, Some(currentTree)),
          currentTree.right, currentTree.parent),
          removedElem)
      } else {
        (Tree(currentTree.elem, EmptyTree, currentTree.right, currentTree.parent),
          removedElem)
      }
    }

    def shortenRightBranch: (Tree[T], T) = {
      val (updatedRightBranch, removedElem) = removeLastBranch(currentTree.right)
      if (updatedRightBranch != EmptyTree) {
        (Tree(currentTree.elem, currentTree.left,
          Tree(updatedRightBranch.elem, updatedRightBranch.left, updatedRightBranch.right, Some(currentTree)),
          currentTree.parent),
          removedElem)
      } else {
        (Tree(currentTree.elem, currentTree.left, EmptyTree, currentTree.parent),
          removedElem)
      }
    }

    if (currentTree.left == EmptyTree && currentTree.right == EmptyTree) {
      return (EmptyTree, currentTree.elem)
    }

    if (currentTree.left.size > currentTree.right.size) shortenLeftBranch else shortenRightBranch
  }

  override def max(): Key = root match {
    case EmptyTree => throw new NullPointerException
    case _ => root.elem
  }

  /**
    * remove the greatest key from the tree
    * which should be root childBranch's value
    *
    * @return
    */
  override def delMax(): Key = {
    val maxValue = max()
    // decrease the size
    N -= 1

    if (N == 0) {
      root = EmptyTree
      last = EmptyTree
    } else {
      val (shortenedTree, removedValue) = removeLastBranch(root)
      root = Tree(removedValue, shortenedTree.left, shortenedTree.right)
      sink(root)
      last = selectNewLast(root)
    }
    maxValue
  }

  def selectNewLast(parent: Tree[T]): Tree[T] = {
    val longestBranch = Tree.longestBranch(parent.left, parent.right)

    if (longestBranch == EmptyTree) {
      return parent
    }

    selectNewLast(longestBranch)
  }

  override def sink(branch: Tree[T]): Unit = {
    if (branch.left == EmptyTree && branch.right == EmptyTree) return

    val greaterChild: Tree[T] = Tree.greaterChild(branch)

    if (greaterChild > branch) {
      swapNodeValues(greaterChild, branch)
      sink(greaterChild)
    }
  }

  def swapNodeValues(childBranch: Tree[Key], parentBranch: Tree[Key]): Unit = {
    val greaterValue = childBranch.elem
    childBranch.asInstanceOf[Branch[Key]].elem = parentBranch.elem
    parentBranch.asInstanceOf[Branch[Key]].elem = greaterValue
  }

  override def swim[U <: Tree[Key]](tree: U): Tree[Key] = {
    tree.parent match {
      case Some(smallerParentBranch) if cmp.lt(smallerParentBranch.elem, tree.elem) =>
        swapNodeValues(tree, smallerParentBranch)
        swim(smallerParentBranch)
      case Some(largerParentBranch) => swim(largerParentBranch)
      // reached root -> return
      case None => tree
    }
  }
}

File: Chapter 2 Sorting/algorithms2_1/src/main/java/ca/vgorcinschi/algorithms1/algorithms2_4/LinkedPQ.java
package ca.vgorcinschi.algorithms1.algorithms2_4;

/**
 * @param <Key>
 * @author Pavel from Tver State University
 */
public class LinkedPQ<Key extends Comparable<Key>> {
    private class Node {
        int N;
        Key data;
        Node parent, left, right;

        public Node(Key data, int N) {
            this.data = data;
            this.N = N;
        }
    }

    // fields
    private Node root;
    private Node lastInserted;

    //helper methods
    private int size(Node x) {
        if (x == null) return 0;
        return x.N;
    }

    private void swim(Node x) {
        if (x == null) return;
        if (x.parent == null) return; // we're at root
        int cmp = x.data.compareTo(x.parent.data);
        if (cmp > 0) {
            swapNodeData(x, x.parent);
            swim(x.parent);
        }
    }

    private void sink(Node nextNode) {
        if (nextNode == null) return;
        Node swapNode;
        if (nextNode.left == null && nextNode.right == null) {
            return;
        } else if (nextNode.left == null) {
            swapNode = nextNode.right;
            int cmp = nextNode.data.compareTo(swapNode.data);
            if (cmp < 0)
                swapNodeData(swapNode, nextNode);
        } else if (nextNode.right == null) {
            swapNode = nextNode.left;
            int cmp = nextNode.data.compareTo(swapNode.data);
            if (cmp < 0)
                swapNodeData(swapNode, nextNode);
        } else {
            int cmp = nextNode.left.data.compareTo(nextNode.right.data);
            if (cmp >= 0) {
                swapNode = nextNode.left;
            } else {
                swapNode = nextNode.right;
            }
            int cmpParChild = nextNode.data.compareTo(swapNode.data);
            if (cmpParChild < 0) {
                swapNodeData(swapNode, nextNode);
                sink(swapNode);
            }
        }
    }

    private void swapNodeData(Node x, Node y) {
        Key temp = x.data;
        x.data = y.data;
        y.data = temp;
    }

    private Node insert(Node x, Key data) {
        if (x == null) {
            lastInserted = new Node(data, 1);
            return lastInserted;
        }
        // compare left and right sizes see where to go
        int leftSize = size(x.left);
        int rightSize = size(x.right);

        if (leftSize <= rightSize) {
            // go to left
            Node inserted = insert(x.left, data);
            x.left = inserted;
            inserted.parent = x;
        } else {
            // go to right
            Node inserted = insert(x.right, data);
            x.right = inserted;
            inserted.parent = x;
        }
        x.N = size(x.left) + size(x.right) + 1;
        return x;
    }

    private Node resetLastInserted(Node x) {
        if (x == null) return null;
        if (x.left == null && x.right == null) return x;
        if (size(x.right) < size(x.left)) return resetLastInserted(x.left);
        else return resetLastInserted(x.right);
    }

    // public methods
    public void insert(Key data) {
        root = insert(root, data);
        swim(lastInserted);
    }

    public Key max() {
        if (root == null) return null;
        return root.data;
    }

    public Key delMax() {
        if (size() == 1) {
            Key ret = root.data;
            root = null;
            return ret;
        }
        swapNodeData(root, lastInserted);
        Node lastInsParent = lastInserted.parent;
        Key lastInsData = lastInserted.data;
        if (lastInserted == lastInsParent.left) {
            lastInsParent.left = null;
        } else {
            lastInsParent.right = null;
        }

        lastInserted = resetLastInserted(root);

        sink(root);

        return lastInsData;
    }

    public int size() {
        return size(root);
    }

    public boolean isEmpty() {
        return size() == 0;
    }
}